### PR TITLE
DE-3885: Implement parsing functions

### PR DESCRIPTION
**Pull request by VladimirShteyn**
## JIRA Ticket
https://tetrascience.atlassian.net/browse/DE-3885

## Description of Changes
Updated the implementation of the parsing functions, mostly by deleting the research implementations. The only major change in logic was improving the error message of `to_boolean`.

## General

- [ ] Not applicable
- [ ] Changelog is added in README.md
  - [ ] Changes are at a high level and do not reference code directly
  - [ ] Scanned for spelling errors (eg. "Code Spell Checker" in VS Code)

### Testing Completed

- [x] Tested manually locally
- [x] Unit tests updated
- [ ] Integration tests updated
- [ ] None (please explain)

### CODE REVIEW

File: task_script_utils/parse.py [hunk 1]
```diff
@@ -190,17 +80,51 @@ def to_boolean(
     :raises ValueError: If the sets overlap or if either set is empty
     """
     if not true_set or not false_set:
-        raise Exception("Fail")
-    if len(true_set & false_set):
-        raise Exception("Fail")
-    true_set_compare = true_set
-    false_set_compare = false_set
+        true_set_quoted_items = [f"'{true_value}'" for true_value in true_set]
```
Comment (mlipschultz) [inline task_script_utils/parse.py:83]: What's the reason for quoting the values for this error?  Could the message for `ValueError` just be:
```python
        raise ValueError(
            f"The set of 'true' values and 'false' values cannot be empty. "
            f"true_set: {true_set}. "
            f"false_set: {false_set}"
        )
```

File: __tests__/unit/parse_functions/test_to_boolean.py [hunk 2]
```diff
@@ -3,30 +3,51 @@
...
-    """Test Empty True Set."""
-    # Arrange
-
-    # Act
-
-    # Assert
-    with pytest.raises(Exception):
-        parse.to_boolean(value, set(), {"False"}, case_sensitive=case_sensitive)
+class TestToBooleanErrors:
+    """Tests for cases where `to_boolean` raises an exception."""
```
Comment (abclay) [inline __tests__/unit/parse_functions/test_to_boolean.py:10]: Is this to make testing in your IDE easier?  I understand the desire to group tests, but I'm not a big fan of having to add `@staticmethod` for all the test functions.

File: pyproject.toml [hunk 3]
```diff
@@ -28,9 +28,7 @@ arrow = "^1.2.2"
 pendulum = "^2.1.2"
 pydash = "^5.1.0"
 numpy = "^1.17.0"
-fastnumbers = "^3.2.1"
 pandas = "1.3.5"
```
Comment (abclay) [inline pyproject.toml:31]: Are numpy and pandas still dependencies in this library?  I don't remember.

File: task_script_utils/parse.py [hunk 4]
```diff
@@ -190,17 +80,51 @@ def to_boolean(
...
+            f"false_set: ({', '.join(false_set_quoted_items)})"
+        )
+
     value_compare = value.strip()
     if not case_sensitive:
-        value_compare = value_compare.lower()
-        true_set_compare = {t.lower() for t in true_set}
-        false_set_compare = {f.lower() for f in false_set}
+        def to_lowercase(string):
+            return string.casefold()
```
Comment (mlipschultz) [inline task_script_utils/parse.py:94]: :trophy:

**Review by abclay — COMMENTED** [review]
Couple quick questions from me.  I will hold off on approving because I wrote the original code and someone else should sign off.

File: task_script_utils/parse.py [hunk 5]
```diff
@@ -190,17 +80,51 @@ def to_boolean(
...
+            false_intersection.update(false_set_compare[value])
+
+        raise ValueError(
+            f"'true_set' and 'false_set' have values in common. "
+            f"true_set: ({', '.join(true_intersection)}). "
+            f"false_set: ({', '.join(false_intersection)})."
+        )
 
+    true_set_compare: Set[str] = set(true_set_compare)
+    false_set_compare: Set[str] = set(false_set_compare)
```
Comment (mlipschultz) [inline task_script_utils/parse.py:127]: Why convert the keys into a set before doing the contains check?

**Review by mlipschultz — CHANGES_REQUESTED** [review]
There's one or two deviations from our agreed-upon plan for v1.  I think those should be fixed.  I also have a couple of other small improvements suggested.

File: pyproject.toml (hunk 3 above)
Comment (mlipschultz) [reply pyproject.toml:31]: Numpy is [currently a dependency](https://github.com/tetrascience/ts-task-script-utils/blob/master/pyproject.toml#L30), but not pandas.

File: .gitignore [hunk 6]
```diff
@@ -13,3 +13,4 @@ dev.*
 .idea/
 venv/
 __pycache__/
+**/.python-version
```
[the remark this answers is not in the document]
Comment (VladimirShteyn) [reply .gitignore:16]: `.python-version` is a file used by `pyenv` to specify the folder's version of Python.

File: pyproject.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (VladimirShteyn) [reply pyproject.toml:31]: Done.

File: task_script_utils/parse.py [hunk 7]
```diff
@@ -1,160 +1,50 @@
...
+    if "_" in value:
+        return None
+    if "e" in value.lower() or "." in value:
         try:
-            array = np.asarray([value], dtype=float)
-            int_val = int(array[0])
-            if int_val == array[0]:
+            float_val = float(value)
+            int_val = int(float_val)
+            if np.isclose(int_val, float_val):
```
[the remark this answers is not in the document]
Comment (VladimirShteyn) [reply task_script_utils/parse.py:24]: In some cases, this implementation is consistent with our earlier decision and in others it's not. Could `float("1")` be evaluated to `1.000000000000001` on some systems, thus erroneously returning `None`? That said, I agree, `None` is the agreed return value when the user passes `to_int("1.000000000000001")`.

File: task_script_utils/parse.py (hunk 1 above)
Comment (VladimirShteyn) [reply task_script_utils/parse.py:83]: Yeah, let's keep it simple. Sometimes it's helpful to visually separate values with quotes, but the commas already do that.

File: task_script_utils/parse.py (hunk 5 above)
Comment (VladimirShteyn) [reply task_script_utils/parse.py:127]: Good point!

File: __tests__/unit/parse_functions/test_to_boolean.py (hunk 2 above)
Comment (VladimirShteyn) [reply __tests__/unit/parse_functions/test_to_boolean.py:10]: Yes, and to make the test names shorter and thus easier to read. I've configured PyCharm to use a test template, so adding `@staticmethod` is no extra work. One can also use a vanilla instance method for tests too, but that's an extra `self` to type. Not sure which is worse.

File: task_script_utils/parse.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (VladimirShteyn) [reply task_script_utils/parse.py]: I changed `math.isclose(int_val, float_val)` to `int_val == float_val`.

[the remark this answers is not in the document]
Comment (mlipschultz) [reply task_script_utils/parse.py]: Did you push that change yet?  I'm still seeing `math.isclose`.

I agree though that we might want to discuss what "close enough" is in a future iteration.

**Review by mlipschultz — APPROVED** [review]

File: pyproject.toml (hunk 3 above)
Comment (VladimirShteyn) [reply pyproject.toml]: I found I could remove `numpy` as a dependency and the tests still pass. 🎉

**Pull request merged**